### ISSUE TITLE
[Deferred] SDL 0002 - Turn by Turn Mobile Managers

### ISSUE THREAD

**Issue by joeljfischer** [smartdevicelink/sdl_evolution#11]
Hello SDL community,

The review of "Turn-by-Turn Mobile Managers" begins now and runs through August 23, 2016. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0002-turn-by-turn-mobile-managers.md
Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/11

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:
- Is the problem being addressed significant enough to warrant a change to SDL?
- Does this proposal fit well with the feel and direction of SDL?
- If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
- How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
  Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Joel Fischer

SDL iOS Maintainer - Livio
joel@livio.io

**Comment by joeljfischer** [comment]
### Review

> Is the problem being addressed significant enough to warrant a change to SDL?

This seems like a worthwhile problem to solve, if it can be solved well. The problem isn't extremely significant, as it is mainly for out-of-date head units and the problem this solves will become less significant over time, but if solved without too much added complexity, it is worthwhile to provide this feature.

> Does this proposal fit well with the feel and direction of SDL?

The solution to this proposal does not fit well with how SDL works. This solution is a very large and complicated effort on the mobile side and would additionally require changes in existing areas of the SDL library as mentioned in the proposal. Below are a list of limitations of this solution:

1. This only works for navigation libraries, not navigation applications. So whichever navigation provider wanted to be a TBT library would have to alter their library or build one if they don't have it. Some navigation libraries don't provide full directions. It would be better if we could somehow integrate with navigation apps themselves (see below for my suggested alternative solution, which takes this approach).

1. This is an OEM based solution based on business deals with nav partners. This limits users' and developers' options. We should not have any OEM based business deals in core SDL (as opposed to branded implementations) if possible.

1. POI apps may need to integrate with multiple navigation libraries, further increasing the size of their applications.

1. There would be UI weirdness, as the POI app would need to not send any UI altering RPCs while this navigation aspect is running. Or we'd need to further complicate the library by preventing developers from sending UI RPCs while the nav SDK has control.

1. This is inherently limited, as they would be dependent on what we have implemented and allowed.

1. This is both outside the spirit of the SDL libraries and complicated to implement technically. The SDL library has never enforced specific UI on developers before now. After carefully reading the design as proposed, this is not as simple as adding a new manager and is likely to generate several new corner cases in addition to requiring the creation of an SDL navigation library interface and an example implementation. We would be committing to upkeep on a good amount of additional technical debt.

> If you have used competitors with a similar feature, how do you feel that this proposal compares to those?

This isn't really applicable, as most competitors only support full map navigation.

> How much effort did you put into your review? A glance, a quick reading, or an in-depth study? Please state explicitly whether you believe that the proposal should be accepted into SDL.

I strongly believe this proposal should not be accepted. I carefully read this proposal and took notes on what it would take to implement into the SDL iOS library.

Another possible solution, one I believe is significantly simpler and more robust would be to enhance SendLocation on the Core side to know what navigation apps are connected and to have Core open a connected navigation app that supports beginning a route to the location. This app would then use its own UI to show the TBT directions as any other app might show UI. The main downside of this approach compared to the current proposal is that the current proposal doesn't require a Core update and perhaps minor RPC changes, the solution I've proposed does. Otherwise, this allows any app the user likes to use have the ability to be the TBT app, in addition to providing custom features that the generic navigation interface would not (such as traffic or road hazards). I believe this solution is more scalable, provides more features, is significantly simpler conceptually, and simpler to implement as well.

A second possible solution would be a more generic interface allowing OEMs to build features like this in extension libraries. There are advantages and disadvantages to a solution like this as well.

**Comment by asm09fsu** [comment]
This definitely feels like a specific implementation that should be done for an internal manager that an app partner would use, not something that should be in SDL.

Also, isn't the point of SDL to not have the user look at their phone? Why are we modifying the Lock screen to show graphics and information when they should not be looking at it? That in and of itself is driver distraction, which is what this proposal is attempting to avoid.

**Comment by joeygrover** [comment]
I strongly agree with Joel's comments. Some additional thoughts:

- TBT is useful for certain regions where there are regulations on using your phone as your main navigation system.
- I would love to see the "services" feature available in Core to redirect requests like this to other connected apps that could handle them, but I don't see that being accomplished in the short term. I disagree that this would be simpler to implement.
- The idea as presented is out of scope in terms of what SDL is and aims to accomplish. 
- The TBT library could be another open source component outside of the main libraries. I see it as a part of a larger "Extensions" project that housed a few different projects, much like the Google play services libs. 

I do not believe this proposal should be accepted.

**Comment by Heath-FoMoCo** [comment]
> The problem isn't extremely significant:

 
This is subjective.  Where are you pulling your information?

> Out of date units


Not true as GEN 1.1 is still be sold as well as non-nav units. This will always happen that non-nav head units are sold.
> problem this solves will become less significant over time:


This is true but I would argue this is true for many features.  It is part of the natural evolution of software.
This may become less significant as mobile navigation is rolled out but may still be relevant for lower end head units.  SendLocation is only functional for embedded nav head units which is low take rate (statistics and be provided offline).
> The solution to this proposal does not fit well with how SDL works.


This is vague.  Do you have examples?  I feel it fits in nicely with the enhanced framework proposals. 
> This solution is a very large and complicated effort on the mobile side


This is subjective: Based on what metrics?
> and would additionally require changes in existing areas of the SDL library as mentioned in the proposal.


When looking from the enhanced framework perspective this proposal requires only minimal changes of the lock screen manager.
> 1.This only works for navigation libraries, not navigation applications. So whichever navigation provider wanted to be a TBT library would have to alter their library or build one if they don't have it. Some navigation libraries don't provide full directions. It would be better if we could somehow integrate with navigation apps themselves (see below for my suggested alternative solution, which takes this approach).


Of course it is a good idea to make navigation apps accessible. But users need to have a navigation app which must be running on a mobile navigation enabled platform. Millions of cars will never be capable of using nav apps. The POI app already ships with the nav feature so it's easiest for users that they don't need to install any additional app. potentially navigation apps are paid apps and therefore may not have a high take rate. Both proposals should coexist.
> 2.This is an OEM based solution based on business deals with nav partners. This limits users' and developers' options. 


Almost every proposal comes from an OEM with business reasons. The proposal was planned to be non-OEM specific. Every OEM can have their own nav library or all of them share one and the same nav library. The planned nav library is capable of being used by multiple OEMs. We don't want to pay for the tbt-library usage on other OEM vehicles. Therefore it's disabled unless OEMs are sharing costs. 

>We should not have any OEM based business deals in core SDL (as opposed to branded implementations) if possible.


This is subjective.  But this solution is architected where SDL can provide a feature (TBT) that costs money that allows an OEM to pay for licensing.  This puts SDL in a more competitive space as SDL will never have a TBT, Voice, or Dictation service.

> 3. POI apps may need to integrate with multiple navigation libraries, further increasing the size of their applications


This is true. The possibility of multiple libraries is more flexible. The ideal case is to share one library with all OEMs (this is covered) which can cover all regions globally (also covered). The proposal is flexible enough if multiple libraries are necessary.


> 4. There would be UI weirdness, as the POI app would need to not send any UI altering RPCs while this navigation aspect is running. Or we'd need to further complicate the library by preventing developers from sending UI RPCs while the nav SDK has control. 


With the enhanced framework the app would get notified that its view got into background which is very similar to the behavior of SendLocation. Anyhow the app should send less/no RPCs anymore within the enhanced framework.


> 5.This is inherently limited, as they would be dependent on what we have implemented and allowed. 


Its also one time effort in SDL instead of n time effort by developers with risk of failures. 

> 6.This is both outside the spirit of the SDL libraries 


Specifically what is it about the spirit?  Is this documented?

> and complicated to implement technically. 


This is subjective.

> The SDL library has never enforced specific UI on developers before now. After carefully reading the design as proposed, this is not as simple as adding a new manager and is likely to generate several new corner cases in addition to requiring the creation of an SDL navigation library interface and an example implementation. 


Instead of generalizing lets talk about specific edge cases.

> We would be committing to upkeep on a good amount of additional technical debt.


This is subjective without data.
>> If you have used competitors with a similar feature, how do you feel that this proposal compares to those?


> This isn't really applicable, as most competitors only support full map navigation.


I don't think this argument is true.  Who are the competitors?  If you mean Google/Apple then they also provide TBT to their developers.
We would rather use full map navigation but in the foreseeable future there is neither SendLocation for mobile navigation apps nor the guarantee for free apps supporting mobile navigation.

>Another possible solution, one I believe is significantly simpler and more robust would be to enhance SendLocation on the Core side to know what navigation apps are connected


This proposal is good but not simpler. There is a proposal in planning but not ready yet because of the complexity with SendLocation for mobile navigation apps. There is no reason to not have both solutions in parallel.

>This definitely feels like a specific implementation that should be done for an internal manager that an app partner would use, not something that should be in SDL.


The proposal is specific to apps offering POI search. Unfortunately there is no way for them to navigate to the destination except SendLocation. This proposal would work for all the POI apps without any change to RPCs, Core etc. and without having a user to install anything additional. It just works.

> Also, isn't the point of SDL to not have the user look at their phone? Why are we modifying the Lock screen to show graphics and information when they should not be looking at it? That in and of itself is driver distraction, which is what this proposal is attempting to avoid.


With this proposal navigation works just with TBT. The moving map on the lock screen is only an additional feature. There is no user interaction possible with the map.


> I would love to see the "services" feature available in Core to redirect requests like this to other connected apps that could handle them, but I don't see that being accomplished in the short term. I disagree that this would be simpler to implement.


Agree. As mentioned already. We are planning a proposal for SendLocation over Core to the mobile nav app which takes more time because of its complexity. The downside of this is the fact that a user needs to have a separate app installed and running.

> The idea as presented is out of scope in terms of what SDL is and aims to accomplish.


Specifically what are you referring to?

> The TBT library could be another open source component outside of the main libraries. I see it as a part of a larger "Extensions" project that housed a few different projects, much like the Google play services libs.


There are two parts already. The TBT library coming from the navigation supplier contains the logic and data. The SDL library should contain the counterpart generating the RPCs out of the data. Therefore the TBT libraries can be seen already as extensions to the TBT hook/dock...

**Comment by joeygrover** [comment]
>> The idea as presented is out of scope in terms of what SDL is and aims to accomplish.

> Specifically what are you referring to?

SmartDeviceLink is an API to allow command and control of apps to other devices. It is not a service provider for additional features apps would like to implement. We don't include things like a media player library to app developers who for some reason want to play music/media. The functionality of their app is a burden of their development. SDL is merely the extension of what their app is already capable of. I'm not opposed to creating separate projects that leverage SDL that we can hand to developers as well, but baking that extra functionality into the main library is not a reflection of the main purpose of SDL. 
 
>> The TBT library could be another open source component outside of the main libraries. I see it as a part of a larger "Extensions" project that housed a few different projects, much like the Google play services libs.

> There are two parts already. The TBT library coming from the navigation supplier contains the logic and data. The SDL library should contain the counterpart generating the RPCs out of the data. Therefore the TBT libraries can be seen already as extensions to the TBT hook/dock...

My point is that there doesn't need to be hooks in the main SDL library. Instead, there could be a TBT-SDL "base" library that either TBT companies can implement in their TBT libraries or it is a stand alone interface that the app developers include. Then either in a constructor or a set method for the 3rd party TBT instance, an app developer leveraging both SDL and a TBT library could pass their proxy instance into the TBT library and the TBT library could leverage SDL. We could create the separate base TBT-SDL library in the open that would handle all the hooks needed for both an app developer and the TBT library while still keeping the main SDL library properly scoped. 


As a comment on the entire review, you mention things as being "subjective" which in some cases you are right, however, the burden of proof or convincing is on the person who proposed the feature or anyone who supports it. Simply saying something is subjective does not fully disqualify the original comment's merritt without a counter argument and proof. Because IMO this process will always be exclusive rather than inclusive in terms of features; only well thought out and necessary ideas will be accepted.

**Comment by joeljfischer** [comment]
This is subjective, that's the whole point of the review. Just pointing out criticisms saying saying, "Subjective" doesn't get us anywhere. The burden of proof here lies on the proposal to justify its implementation, not on the reviewers to justify their criticisms. The proposal, if it wishes to justify itself, should be providing numbers, not criticizing reviewers for not doing so. Every single thing we add creates technical debt. We should, therefore, default to *not* implementing a new feature unless it truly justifies itself by having a good cost-benefit analysis where the cost (in terms of both technical debt and developer time to implement) does not outweigh the benefits to users. I strongly do not believe this proposal is at that threshold. Of course that's subjective, and you can say why you *do* believe it is within that threshold and we can then compare the pros and cons.

You commonly seems to say that my criticism that this will be heavy on technical debt and difficult to implement well is "subjective." Again, of course it is, and your opinion that it is not is subjective. I'm not going to spend the next 2 months implementing this proposal to prove it's difficult to implement. The review team will have to judge how much my word is worth on the subject. However, nobody knows the iOS library better than I do, and I'm quite confident that this will be a headache and is not easy at all. I don't have the space or the time to prove this, but to point out one specific misunderstanding:

> With the enhanced framework the app would get notified that its view got into background which is very similar to the behavior of SendLocation. Anyhow the app should send less/no RPCs anymore within the enhanced framework.

This is simply not how the 4.3 dev api (EF) works at this point. Artificially disabling the developer's abilities to show functionality is not at all how SDL has ever worked and would require a large amount of work to implement.

This idea is out of scope of SDL because we do not, and never have done *anything* like this. We have always focussed on the connection between an app and the head unit. This proposal would broaden the scope of SDL to create an SDL / app / libraries circle of data in which SDL now controls the app's lifecycle on the head unit (as opposed to it strictly being controlled by the head unit) and communicating with an additional vendored library to provide additional functionality. I'm not totally opposed to this, though I would prefer to see a more generic form of extensions rather than specific feature based extensions like this.

Finally, the idea that we should implement this functionality twice is, frankly, absurd. Why in the world would we implement two wholly different ways to achieve the same result? That's just confusing. I still strongly believe the other idea to have SendLocation work with other mobile apps is the better solution. It is objectively *simpler* in terms of having less moving parts, it's just not necessarily *easier*. The SendLocation solution fits better with how SDL works now, is more predictable, and I believe it's easier for developers to implement (they just become SDL enabled rather than needing a library that can be extended into SDL). Yes, users need another app, but it could be the apps they already use rather than nav services they don't like or trust, and their apps will not get bloated by needing to implement SDL + possible multiple nav partners. Plus, it's free for OEMs.

Finally, I believe my criticism that this should _at least_ be deferred until post v5.0 on iOS stands, though I would prefer to see outright rejection.

**Comment by joeljfischer** [comment]
This proposal has been deferred and will be revisited by the review team at a later time.

**Comment by Heath-FoMoCo** [comment]
> This is subjective, that's the whole point of the review. Just pointing out criticisms saying saying, "Subjective" doesn't get us anywhere. The burden of proof here lies on the proposal to justify its implementation, not on the reviewers to justify their criticisms. The proposal, if it wishes to justify itself, should be providing numbers, not criticizing reviewers for not doing so. Every single thing we add creates technical debt. We should, therefore, default to not implementing a new feature unless it truly justifies itself by having a good cost-benefit analysis where the cost (in terms of both technical debt and developer time to implement) does not outweigh the benefits to users. I strongly do not believe this proposal is at that threshold. Of course that's subjective, and you can say why you do believe it is within that threshold and we can then compare the pros and cons.

What you say here is in complete contrast to what you posted here: 

> The problem isn't extremely significant, as it is mainly for out-of-date head units and the problem this solves will become less significant over time, but if solved without too much added complexity, it is worthwhile to provide this feature.

There is no data in that post.

**Comment by joeygrover** [comment]
Here is a high level for chart of how extensions might work in SDL as well as who will maintain them. 

- Extensions beyond TBT are simply used as example extensions and do not denote any current plans to include them. 
- For the sake of brevity, not all enhanced framework managers are shown.
- Although not explicitly shown, the extension manager will communicate with other enhanced framework managers.
- Extension creators are the developers/groups who come up with and create the extension library of that type. This library will interact with  the SDL library as well as one or many service provider for that type of extension.
- Extension providers are the developers/groups that create a service library that will interface with the extension library of their service type. 
- The extension base interface will be designed to work directly with the extension manager. The interface is abstract and must be implemented by an actual extension type. It is not a standalone product.

[<img src="http://i.imgur.com/dyZYGUf.png">](http://i.imgur.com/dyZYGUf.png)

**Comment by theresalech** [comment]
The committee agreed to reject this proposal, as it was decided this should be implemented via generic mobile extensions, and a separate proposal will be written for that solution.
